### PR TITLE
Ensure propertyChanged event occurs for setState and setStateLabel

### PR DESCRIPTION
**Pull request by tlambert03**
Akin to #38, when `core.setState()` is called it will _sometimes_ cause a `core.events.propertyChanged` event to be emitted, in a device-dependent manor.  (for example, in the demo config, it works for `Objective`, but not for `Dichroic`).  This pulls the logic from #38 into a more generally useful context manager and uses it for `setState` and `setStateLabel` as well

### CODE REVIEW

File: pymmcore_plus/core/_mmcore_plus.py [hunk 1]
```diff
@@ -111,38 +112,25 @@ def __del__(self):
...
-        ):  # block the native event.
-            super().setProperty(label, propName, propValue)
-        after = super().getProperty(label, propName)
-        if before != after:
-            self.events.propertyChanged.emit(label, propName, after)
+    @synchronized(lock)
+    def setStateLabel(self, stateDeviceLabel: str, stateLabel: str) -> None:
+        """Set state (by label) on stateDeviceLabel, with reliable event emission."""
+        with self._property_change_emission_ensured(
+            stateDeviceLabel, ("State", "Label")
```
Comment (ianhi) [inline pymmcore_plus/core/_mmcore_plus.py:131]: What happens if a state doesn't have an associated label? I suspect that a blank string will be sent, but we should check that this doesn't cause an error before merging.

File: pymmcore_plus/_tests/test_core.py [hunk 2]
```diff
@@ -353,6 +353,28 @@ def test_set_property_events(core: CMMCorePlus):
...
 
+def test_set_state_events(core: CMMCorePlus):
+    mock = MagicMock()
+    core.events.propertyChanged.connect(mock)
+    assert core.getState("Objective") == 1
+    core.setState("Objective", 3)
+    mock.assert_has_calls(
+        [
+            call("Objective", "State", "3"),
+            call("Objective", "Label", "Nikon 20X Plan Fluor ELWD"),
```
Comment (ianhi) [inline pymmcore_plus/_tests/test_core.py:364]: Mentioning to expect both of these and only handle one of them would be great to include in a `working with properties and state devices` example.

**Review by ianhi — COMMENTED** [review]
One concern about what happens for a state device with no labels - which I can check. Otherwise Looks good!

**Comment by tlambert03** [comment]
Good points all around!  Thanks, will address tomorrow.

File: pymmcore_plus/core/_mmcore_plus.py (hunk 1 above)
Comment (ianhi) [reply pymmcore_plus/core/_mmcore_plus.py:131]: Looks like they they default to `State-1` and all is well

**Review by ianhi — APPROVED** [review]
No longer concerns about unlabelled states. So this is good to go from my perspective.

Suggestions about docstring + example stand though.

**Pull request merged**